Serialization Errors Cause Flaky UI Behaviour

**Error description:**

When an error occurs during session serialization, the UI often displays "Offline" / "Online" notifications on the client side that are not related to the actual state of the connection to the server side.

It is possible to use the UI, but sometimes certain operations randomly do not work as though the session was not available, even though it should be.


**Expected behaviour:**

Errors during session serialization should not impair users using the app.
Session serialization errors should not cause other errors in the app.


**Details:**

We are using Spring Boot + Redis to store the serialized sessions.
There was an error in Redis where the key expiration was not configured correctly and an Out-Of-Memory error was caused.
This led to the following exception in our Vaadin application:

![image (1)](https://github.com/vaadin/kubernetes-kit/assets/144342931/731ed735-a5fd-4a4f-acb0-ee16d7fd9dfc)

```
Servlet.service() for servlet [springServlet] in context with path [] threw exception
org.springframework.data.redis.RedisSystemException: Error in execution
	at org.springframework.data.redis.connection.lettuce.LettuceExceptionConverter.convert(LettuceExceptionConverter.java:52)
	at org.springframework.data.redis.connection.lettuce.LettuceExceptionConverter.convert(LettuceExceptionConverter.java:50)
	at org.springframework.data.redis.connection.lettuce.LettuceExceptionConverter.convert(LettuceExceptionConverter.java:41)
	at org.springframework.data.redis.PassThroughExceptionTranslationStrategy.translate(PassThroughExceptionTranslationStrategy.java:40)
	at org.springframework.data.redis.FallbackExceptionTranslationStrategy.translate(FallbackExceptionTranslationStrategy.java:38)
	at org.springframework.data.redis.connection.lettuce.LettuceConnection.convertLettuceAccessException(LettuceConnection.java:306)
	at org.springframework.data.redis.connection.lettuce.LettuceConnection.await(LettuceConnection.java:1010)
	at org.springframework.data.redis.connection.lettuce.LettuceConnection.lambda$doInvoke$3(LettuceConnection.java:443)
	at org.springframework.data.redis.connection.lettuce.LettuceInvoker$Synchronizer.invoke(LettuceInvoker.java:673)
	at org.springframework.data.redis.connection.lettuce.LettuceInvoker$DefaultSingleInvocationSpec.get(LettuceInvoker.java:589)
	at org.springframework.data.redis.connection.lettuce.LettuceStringCommands.set(LettuceStringCommands.java:102)
	at org.springframework.data.redis.connection.DefaultedRedisConnection.set(DefaultedRedisConnection.java:319)
	at com.vaadin.kubernetes.starter.sessiontracker.backend.RedisConnector.markSerializationStarted(RedisConnector.java:76)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.queueSerialization(SessionSerializer.java:225)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.serialize(SessionSerializer.java:185)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.serialize(SessionSerializer.java:171)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionTrackerFilter.doFilter(SessionTrackerFilter.java:78)
	at jakarta.servlet.http.HttpFilter.doFilter(HttpFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
[...]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.lettuce.core.RedisCommandExecutionException: OOM command not allowed when used memory > 'maxmemory'.
	at io.lettuce.core.internal.ExceptionFactory.createExecutionException(ExceptionFactory.java:147)
[...]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 common frames omitted
```

Which caused this follow-up error:

![14-03-2024_09-23-53](https://github.com/vaadin/kubernetes-kit/assets/144342931/8ec6f2e6-5379-4341-8fa1-f17a8706b73f)

```
Servlet.service() for servlet [springServlet] in context with path [] threw exception
org.springframework.dao.QueryTimeoutException: Redis command timed out
	at org.springframework.data.redis.connection.lettuce.LettuceExceptionConverter.convert(LettuceExceptionConverter.java:68)
	at org.springframework.data.redis.connection.lettuce.LettuceExceptionConverter.convert(LettuceExceptionConverter.java:41)
	at org.springframework.data.redis.PassThroughExceptionTranslationStrategy.translate(PassThroughExceptionTranslationStrategy.java:40)
	at org.springframework.data.redis.FallbackExceptionTranslationStrategy.translate(FallbackExceptionTranslationStrategy.java:38)
	at org.springframework.data.redis.connection.lettuce.LettuceConnection.convertLettuceAccessException(LettuceConnection.java:306)
	at org.springframework.data.redis.connection.lettuce.LettuceConnection.await(LettuceConnection.java:1010)
	at org.springframework.data.redis.connection.lettuce.LettuceConnection.lambda$doInvoke$3(LettuceConnection.java:443)
	at org.springframework.data.redis.connection.lettuce.LettuceInvoker$Synchronizer.invoke(LettuceInvoker.java:673)
	at org.springframework.data.redis.connection.lettuce.LettuceInvoker$DefaultSingleInvocationSpec.get(LettuceInvoker.java:589)
	at org.springframework.data.redis.connection.lettuce.LettuceStringCommands.set(LettuceStringCommands.java:102)
	at org.springframework.data.redis.connection.DefaultedRedisConnection.set(DefaultedRedisConnection.java:319)
	at com.vaadin.kubernetes.starter.sessiontracker.backend.RedisConnector.markSerializationStarted(RedisConnector.java:76)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.queueSerialization(SessionSerializer.java:225)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.serialize(SessionSerializer.java:185)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionSerializer.serialize(SessionSerializer.java:171)
	at com.vaadin.kubernetes.starter.sessiontracker.SessionTrackerFilter.doFilter(SessionTrackerFilter.java:78)
	at jakarta.servlet.http.HttpFilter.doFilter(HttpFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
[...]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.lettuce.core.RedisCommandTimeoutException: Command timed out after 5 second(s)
	at io.lettuce.core.internal.ExceptionFactory.createTimeoutException(ExceptionFactory.java:59)
	at io.lettuce.core.internal.Futures.awaitOrCancel(Futures.java:246)
	at io.lettuce.core.LettuceFutures.awaitOrCancel(LettuceFutures.java:74)
	at org.springframework.data.redis.connection.lettuce.LettuceConnection.await(LettuceConnection.java:1008)
	... 152 common frames omitted
```


The moment when the Redis memory was filled with keys:
![14-03-2024_09-24-32](https://github.com/vaadin/kubernetes-kit/assets/144342931/11f37074-13dc-4a11-81ff-613488df20c2)


During the time it took us to solve the Redis error, the Vaadin application UI behaved very strangely: It was constantly showing "Offline" / "Online" notifications. We were suspecting that the session serialization interfered with the Vaadin push system.

--- a/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializer.java
+++ b/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializer.java
@@ -19,6 +19,7 @@ import java.util.HashSet;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
@@ -229,17 +230,30 @@ public class SessionSerializer
         getLogger().debug(
                 "Starting asynchronous serialization of session {} with distributed key {}",
                 sessionId, clusterKey);
-        backendConnector.markSerializationStarted(clusterKey);
         pending.put(sessionId, true);
-
-        executorService.submit(() -> {
-            Consumer<SessionInfo> whenSerialized = sessionInfo -> {
-                backendConnector.sendSession(sessionInfo);
-                backendConnector.markSerializationComplete(clusterKey);
-            };
-
-            handleSessionSerialization(sessionId, attributes, whenSerialized);
-        });
+        // Backend operations are performed asynchronously to prevent the UI to
+        // freeze in case of errors, timeouts or slow performance.
+        // Current session is immediately marked as 'serialization pending',
+        // because if 'markSerializationStarted' is hanging, it does not make
+        // sense to retry the operation instantly.
+        CompletableFuture.runAsync(
+                () -> backendConnector.markSerializationStarted(clusterKey),
+                executorService).whenComplete((unused, error) -> {
+                    if (error != null) {
+                        pending.remove(sessionId);
+                        getLogger().debug(
+                                "Failed marking serialization start for of session {} with distributed key {}",
+                                sessionId, clusterKey, error);
+                    } else {
+                        Consumer<SessionInfo> whenSerialized = sessionInfo -> {
+                            backendConnector.sendSession(sessionInfo);
+                            backendConnector
+                                    .markSerializationComplete(clusterKey);
+                        };
+                        handleSessionSerialization(sessionId, attributes,
+                                whenSerialized);
+                    }
+                });
     }
 
     private void handleSessionSerialization(String sessionId,

--- a/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/serialization/debug/DebugBackendConnector.java
+++ b/kubernetes-kit-starter/src/main/java/com/vaadin/kubernetes/starter/sessiontracker/serialization/debug/DebugBackendConnector.java
@@ -33,7 +33,7 @@ class DebugBackendConnector implements BackendConnector {
 
     private SessionInfo sessionInfo;
 
-    private CountDownLatch latch;
+    private CountDownLatch latch = new CountDownLatch(2);
 
     @Override
     public void sendSession(SessionInfo sessionInfo) {
@@ -47,7 +47,7 @@ class DebugBackendConnector implements BackendConnector {
 
     @Override
     public void markSerializationStarted(String clusterKey) {
-        latch = new CountDownLatch(1);
+        latch.countDown();
         job.serializationStarted();
     }
 

--- a/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
+++ b/kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java
@@ -1,12 +1,14 @@
 package com.vaadin.kubernetes.starter.sessiontracker;
 
 import jakarta.servlet.http.HttpSession;
+
 import java.io.IOException;
 import java.io.Serializable;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.locks.Lock;
@@ -31,6 +33,8 @@ import com.vaadin.kubernetes.starter.sessiontracker.backend.SessionInfo;
 import com.vaadin.kubernetes.starter.sessiontracker.serialization.TransientHandler;
 
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.fail;
 import static org.mockito.ArgumentMatchers.any;
@@ -82,6 +86,9 @@ class SessionSerializerTest {
 
     @Test
     void serialize_optimisticLocking_sessionNotLocked() {
+        AtomicBoolean serializationStarted = new AtomicBoolean();
+        doAnswer(i -> serializationStarted.getAndSet(true)).when(connector)
+                .markSerializationStarted(clusterSID);
         AtomicBoolean serializationCompleted = new AtomicBoolean();
         doAnswer(i -> serializationCompleted.getAndSet(true)).when(connector)
                 .markSerializationComplete(clusterSID);
@@ -89,6 +96,7 @@ class SessionSerializerTest {
         vaadinSession.setLockTimestamps(10, 20);
 
         serializer.serialize(httpSession);
+        await().during(100, MILLISECONDS).untilTrue(serializationStarted);
         verify(connector).markSerializationStarted(clusterSID);
 
         await().atMost(1000, MILLISECONDS).untilTrue(serializationCompleted);
@@ -97,6 +105,9 @@ class SessionSerializerTest {
 
     @Test
     void serialize_optimisticLocking_sessionLocked() {
+        AtomicBoolean serializationStarted = new AtomicBoolean();
+        doAnswer(i -> serializationStarted.getAndSet(true)).when(connector)
+                .markSerializationStarted(clusterSID);
         AtomicBoolean serializationCompleted = new AtomicBoolean();
         doAnswer(i -> serializationCompleted.getAndSet(true)).when(connector)
                 .markSerializationComplete(clusterSID);
@@ -105,6 +116,7 @@ class SessionSerializerTest {
         vaadinSession.setLockTimestamps(30, 20);
 
         serializer.serialize(httpSession);
+        await().during(100, MILLISECONDS).untilTrue(serializationStarted);
         verify(connector).markSerializationStarted(clusterSID);
 
         await().during(100, MILLISECONDS).untilFalse(serializationCompleted);
@@ -118,6 +130,9 @@ class SessionSerializerTest {
 
     @Test
     void serialize_optimisticLocking_sessionChanged() {
+        AtomicBoolean serializationStarted = new AtomicBoolean();
+        doAnswer(i -> serializationStarted.getAndSet(true)).when(connector)
+                .markSerializationStarted(clusterSID);
         AtomicBoolean serializationCompleted = new AtomicBoolean();
         doAnswer(i -> serializationCompleted.getAndSet(true)).when(connector)
                 .markSerializationComplete(clusterSID);
@@ -128,6 +143,7 @@ class SessionSerializerTest {
         httpSession.setAttribute("DELAY", new SerializationDelay(300));
 
         serializer.serialize(httpSession);
+        await().during(100, MILLISECONDS).untilTrue(serializationStarted);
         verify(connector).markSerializationStarted(clusterSID);
 
         await().during(100, MILLISECONDS).untilFalse(serializationCompleted);
@@ -174,6 +190,9 @@ class SessionSerializerTest {
 
     @Test
     void serialize_pessimisticLocking() {
+        AtomicBoolean serializationStarted = new AtomicBoolean();
+        doAnswer(i -> serializationStarted.getAndSet(true)).when(connector)
+                .markSerializationStarted(clusterSID);
         AtomicBoolean serializationCompleted = new AtomicBoolean();
         doAnswer(i -> serializationCompleted.getAndSet(true)).when(connector)
                 .markSerializationComplete(clusterSID);
@@ -183,6 +202,7 @@ class SessionSerializerTest {
         vaadinSession.setLockTimestamps(30, 20);
 
         serializer.serialize(httpSession);
+        await().during(100, MILLISECONDS).untilTrue(serializationStarted);
         verify(connector).markSerializationStarted(clusterSID);
 
         await().during(TEST_OPTIMISTIC_SERIALIZATION_TIMEOUT_MS + 100,
@@ -243,6 +263,10 @@ class SessionSerializerTest {
 
     @Test
     void serialize_notSerializableException_notFallbackToPessimistic() {
+        AtomicBoolean serializationStarted = new AtomicBoolean();
+        doAnswer(i -> serializationStarted.getAndSet(true)).when(connector)
+                .markSerializationStarted(clusterSID);
+
         AtomicBoolean serializationCompleted = new AtomicBoolean();
         doAnswer(i -> serializationCompleted.getAndSet(true)).when(connector)
                 .markSerializationComplete(clusterSID);
@@ -263,12 +287,117 @@ class SessionSerializerTest {
         }
 
         serializer.serialize(httpSession);
+        await().atMost(100, MILLISECONDS).untilTrue(serializationStarted);
         verify(connector).markSerializationStarted(clusterSID);
 
         await().atMost(500, MILLISECONDS).untilTrue(serializationCompleted);
         verify(connector).sendSession(isNull());
         locks.forEach(l -> verify(l, never()).lock());
 
+    }
+
+    @Test
+    void serialize_slowBackendConnector_shouldNotBlockExecution()
+            throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(2);
+        doAnswer(i -> {
+            // Simulate slow backed
+            Thread.sleep(1000);
+            latch.countDown();
+            return null;
+        }).when(connector).markSerializationStarted(anyString());
+        doAnswer(i -> {
+            latch.countDown();
+            return null;
+        }).when(connector).markSerializationComplete(anyString());
+
+        long start = System.nanoTime();
+        serializer.serialize(httpSession);
+        long elapsedNanos = System.nanoTime() - start;
+        Assertions.assertTrue(elapsedNanos < MILLISECONDS.toNanos(100),
+                "Execution should not be blocked by slow backend connector");
+
+        latch.await();
+        verify(connector).markSerializationStarted(anyString());
+        verify(connector).sendSession(notNull());
+        verify(connector).markSerializationComplete(anyString());
+    }
+
+    @Test
+    void serialize_slowBackendConnector_additionalRequestShouldNotBeEnqueued()
+            throws InterruptedException {
+        CountDownLatch serializationStarted = new CountDownLatch(2);
+        doAnswer(i -> {
+            // Simulate slow backed
+            Thread.sleep(1000);
+            serializationStarted.countDown();
+            return null;
+        }).when(connector).markSerializationStarted(anyString());
+        CountDownLatch serializationCompleted = new CountDownLatch(2);
+        doAnswer(i -> {
+            // Simulate slow backed
+            Thread.sleep(100);
+            serializationCompleted.countDown();
+            return null;
+        }).when(connector).markSerializationComplete(anyString());
+
+        long start = System.nanoTime();
+        serializer.serialize(httpSession);
+        long elapsedNanos = System.nanoTime() - start;
+        Assertions.assertTrue(elapsedNanos < MILLISECONDS.toNanos(100),
+                "Execution should not be blocked by slow backend connector (Took "
+                        + NANOSECONDS.toMillis(elapsedNanos) + " ms)");
+
+        // Serialization pending, should not enqueue
+        serializer.serialize(httpSession);
+        serializer.serialize(httpSession);
+        serializer.serialize(httpSession);
+        serializer.serialize(httpSession);
+
+        await().until(() -> serializationStarted.getCount() == 1);
+
+        await().until(() -> serializationCompleted.getCount() == 1);
+        // Not pending anymore, should enqueue again
+        start = System.nanoTime();
+        serializer.serialize(httpSession);
+        elapsedNanos = System.nanoTime() - start;
+        Assertions.assertTrue(elapsedNanos < MILLISECONDS.toNanos(100),
+                "Execution should not be blocked by slow backend connector (Took "
+                        + NANOSECONDS.toMillis(elapsedNanos) + " ms)");
+        serializationStarted.await();
+
+        serializationCompleted.await();
+        verify(connector, times(2)).markSerializationStarted(anyString());
+        verify(connector, times(2)).sendSession(notNull());
+        verify(connector, times(2)).markSerializationComplete(anyString());
+    }
+
+    @Test
+    void serialize_backendConnectorFailureOnStart_shouldNotFail()
+            throws InterruptedException {
+        AtomicBoolean doFail = new AtomicBoolean(true);
+        CountDownLatch latch = new CountDownLatch(2);
+        doAnswer(i -> {
+            try {
+                if (doFail.get()) {
+                    throw new RuntimeException("BOOM!");
+                }
+                return null;
+            } finally {
+                latch.countDown();
+            }
+        }).when(connector).markSerializationStarted(anyString());
+
+        Assertions.assertDoesNotThrow(() -> serializer.serialize(httpSession),
+                "Backend connector failure on start should not be propagated");
+        await().until(() -> latch.getCount() == 1);
+
+        // Second call should work
+        doFail.set(false);
+        serializer.serialize(httpSession);
+
+        latch.await();
+        verify(connector, times(2)).markSerializationStarted(anyString());
     }
 
     @Test

